### DISCUSSION TITLE
What is the purpose of `app:created` hook, which navigates to `resolvedInitialRoute` after all plugins finished their work, in `nuxt:router` plugin?

### DISCUSSION THREAD

**Discussion by nicky1038** [nuxt/nuxt#27979] [Questions]
Hi everyone!

I've struggled into a need of using `navigateTo` to perform some internal app navigation inside my nuxt plugin. It is a part of app initialization process and logically it should not be performed in middlewares but in plugin.

But it turns out that `navigateTo` may be useless if it is performed inside nuxt plugin, because a bit earlier `nuxt:router` plugin remembers the initial URL (`resolvedInitialRoute`) and registers `app:created` hook that performs navigation to that URL after all plugins finished their work. Thus my `navigateTo` call has no effect.

Could anyone please clarify what this `app:created` hook is used for in `nuxt:router` plugin? At a glance, it just doesn't allow me to achieve what I need. I could not find any relevant information anywhere else.

**Comment by nicky1038** [comment]
@danielroe Hi! 
Sorry for disturbing you.
Could you please clarify why exactly do you force an app to navigate to its initial url in  `app:created` hook in [this commit](https://github.com/nuxt/nuxt/commit/4826918ed09f16a305dd9989a185204a81899ccc)?
Because of this, there is no option to make any internal navigation via `navigateTo` in nuxt plugins. It is a very strong but unclear limitation.

**Reply by danielroe** [reply]
Because otherwise user middleware that is registered in a plugin would never run on the initial load.

You can perform navigation in a plugin by registering a custom middleware or a router hook.

**Reply by nicky1038** [reply]
> Because otherwise user middleware that is registered in a plugin would never run on the initial load.

So, is it true that the purpose of this navigation call is just to trigger vue-router to call all middlewares? If so, why do we navigate specifically to the initial route, thus ignoring navigation calls made by plugins? For example, if a plugin makes internal redirect from `/` to `/a`, and then in `app:created` hook nuxt could make a redirect to `/a`, and not to `/`, all middlewares would be called too.

> You can perform navigation in a plugin by registering a custom middleware or a router hook.

Unfortunately, this solution doesn't suit for me. I have a nuxt plugin that can be used in various projects which are not mine. It performs some initialization logic and in certain cases makes redirects. Also it is needed to be ran before any project code. Since projects can have their own plugins, and my code should be executed before them, my init logic (including redirects) should be also ran in a plugin. The only reason this plugin doesn't work properly is that nuxt navigates to the initial url in `app:created` hook, instead of the new route. That is why I am interested why nuxt behaves exactly this way and if we can change this url.

**Reply by danielroe** [reply]
If you perform a navigation in `app:created` on the server, then by the time it loads on the client it should already be the correct path, right? So this is only if you perform a navigation _only_ on client-side, such as with `ssr: false` or some other situation.

I think we might be able to resolve this in any case - would you create an issue with reproduction so we can track it? 🙏

**Reply by nicky1038** [reply]
Indeed, the sandbox I have tested this case in has `ssr: false` flag set up in nuxt config.
Sure thing, I will create an issue with a small reproduction demo, thank you for your replies :)

**Reply by nicky1038** [reply]
Done 👌
https://github.com/nuxt/nuxt/issues/28174
